src/Main.hs

```diff
@@ -59,21 +63,44 @@ instance FromJSON DeckPost where
 
 type FlashCardAPI = "decks" :> Get '[JSON] [Entity Deck]
                   :<|> "decks" :> ReqBody '[JSON] DeckPost :> Post '[JSON] (Entity Deck)
-                  :<|> "decks" :> Capture "id" Int64 :> Get '[JSON] (Entity Deck)
+                  :<|> "decks" :> Capture "id" Int64 :> Throws (NotFoundError Deck) :> Get '[JSON] (Entity Deck)
 
-server :: Server FlashCardAPI
+data NotFoundError a = NotFoundError Int64 deriving (Eq, Read, Show)
+
+instance ErrStatus (NotFoundError a) where
+  toErrStatus :: NotFoundError a -> Status
+  toErrStatus _ = status404
+
+instance ToJSON (NotFoundError Deck) where
+  toJSON :: (NotFoundError Deck) -> Value
+  toJSON (NotFoundError deckId) = toJSON $ "no Deck found with id: " ++ show deckId
+
+allDecksH :: Handler [(Entity Deck)]
+allDecksH = liftIO getDecks
+
+postDecksH :: DeckPost -> Handler (Entity Deck)
+postDecksH deck = liftIO $ insertDeck deck
+
+getDeckByIdH :: Int64 -> Handler (Envelope '[(NotFoundError Deck)] (Entity Deck))
+getDeckByIdH deckId = liftIO $ do
+  mdeck <- getDeckById deckId
+  case mdeck of
+    Nothing -> pureErrEnvelope ((NotFoundError deckId) :: (NotFoundError Deck))
+    Just deck -> pureSuccEnvelope deck
+
+server :: ServerT FlashCardAPI Handler
 server = allDecksH  :<|> postDecksH :<|> getDeckByIdH
-  where allDecksH = liftIO getDecks
-        postDecksH deck = liftIO $ insertDeck deck
-        getDeckByIdH deckId = liftIO $ getDeckById deckId
+
+asSqlBackendReader :: ReaderT SqlBackend m a -> ReaderT SqlBackend m a
```

Comment:
Man I ended up using this type hitching trick too. I even used the same name -- we must be googling the same sources.